File: alpha_engine/evidence.py
import html
import logging
import urllib.parse
import xml.etree.ElementTree as ET

import httpx

from .config import settings

log = logging.getLogger(__name__)


def current_evidence(query: str) -> list[dict[str, str]]:
    """Fetch current Google News RSS evidence without requiring another API key."""
    if not settings.web_research_enabled:
        return []
    url = 'https://news.google.com/rss/search?' + urllib.parse.urlencode({
        'q': query,
        'hl': 'en-US',
        'gl': 'US',
        'ceid': 'US:en',
    })
    try:
        response = httpx.get(url, timeout=settings.web_research_timeout, follow_redirects=True,
                             headers={'User-Agent': 'polymarket-alpha-engine/1.1'})
        response.raise_for_status()
        root = ET.fromstring(response.content)
        items = []
        for item in root.findall('.//item')[:settings.web_research_max_items]:
            title = html.unescape(item.findtext('title') or '').strip()
            link = (item.findtext('link') or '').strip()
            published = (item.findtext('pubDate') or '').strip()
            source = item.find('source')
            source_name = source.text.strip() if source is not None and source.text else ''
            if title:
                items.append({'title': title, 'source': source_name, 'published': published, 'url': link})
        return items
    except Exception as exc:
        log.warning('web evidence unavailable for %r: %s', query, exc)
        return []


def format_evidence(items: list[dict[str, str]]) -> str:
    if not items:
        return 'No current web evidence could be retrieved. Treat the forecast as stale and lower confidence.'
    lines = []
    for index, item in enumerate(items, 1):
        lines.append(f"{index}. {item['title']} | {item['source']} | {item['published']} | {item['url']}")
    return '\n'.join(lines)

File: worker/src/evidence.ts
import type { Settings } from "./config";
import type { EvidenceItem } from "./types";

const FORMULAIC_WORDS = new Set([
  "will", "would", "does", "do", "did", "is", "are", "be", "before", "after",
  "by", "on", "in", "of", "the", "a", "an", "this", "that", "market",
  "resolve", "resolves", "yes", "no",
]);

export function buildNewsQuery(question: string): string {
  const words = question.match(/[A-Za-z0-9][A-Za-z0-9'&.-]*/g) || [];
  const useful: string[] = [];
  for (const word of words) {
    const normalized = word.toLowerCase().replace(/^[.-]+|[.-]+$/g, "");
    if (FORMULAIC_WORDS.has(normalized)) continue;
    if (/^20\d{2}$/.test(normalized)) continue;
    useful.push(word);
  }
  const compact = useful.slice(0, 12).join(" ").trim();
  return compact ? `${compact} when:30d` : `"${question.trim()}" when:30d`;
}

function unescapeHtml(text: string): string {
  return text
    .replace(/<!\[CDATA\[([\s\S]*?)\]\]>/g, "$1")
    .replace(/&amp;/g, "&")
    .replace(/&lt;/g, "<")
    .replace(/&gt;/g, ">")
    .replace(/&quot;/g, '"')
    .replace(/&#39;/g, "'")
    .replace(/&apos;/g, "'")
    .trim();
}

function extractTag(block: string, tag: string): string {
  const match = block.match(new RegExp(`<${tag}[^>]*>([\\s\\S]*?)</${tag}>`, "i"));
  return match ? unescapeHtml(match[1]) : "";
}

export async function currentEvidence(
  settings: Settings,
  question: string,
): Promise<EvidenceItem[]> {
  if (!settings.webResearchEnabled) return [];

  const query = buildNewsQuery(question);
  const url = new URL("https://news.google.com/rss/search");
  url.searchParams.set("q", query);
  url.searchParams.set("hl", "en-US");
  url.searchParams.set("gl", "US");
  url.searchParams.set("ceid", "US:en");

  try {
    const response = await fetch(url.toString(), {
      headers: { "User-Agent": "polymarket-alpha-engine-worker/1.1" },
      signal: AbortSignal.timeout(settings.webResearchTimeoutMs),
    });
    if (!response.ok) throw new Error(`Google News RSS request failed: ${response.status}`);
    const xml = await response.text();

    const items: EvidenceItem[] = [];
    const itemBlocks = xml.match(/<item>[\s\S]*?<\/item>/g) || [];
    for (const block of itemBlocks.slice(0, settings.webResearchMaxItems)) {
      const title = extractTag(block, "title");
      const link = extractTag(block, "link");
      const published = extractTag(block, "pubDate");
      const source = extractTag(block, "source");
      if (title) {
        items.push({ title, source, published, url: link, query });
      }
    }
    return items;
  } catch {
    return [];
  }
}

export function formatEvidence(items: EvidenceItem[]): string {
  if (items.length === 0) {
    return "No current web evidence could be retrieved. Treat the forecast as stale and lower confidence.";
  }
  return items
    .map((item, index) => `${index + 1}. ${item.title} | ${item.source} | ${item.published} | ${item.url}`)
    .join("\n");
}
